File: ra-idp/src/administration/usecases/users.test.ts
/**
 * Layer 3 — Application Logic (administration use cases)
 *
 * 認可境界 (RBAC / CSRF / session) を持たない純粋な use case 層を確認する。
 * HTTP 統合は adapters/http/admin-user-routes.test.ts で別途見る。
 */

import { describe, expect, it } from 'bun:test'

import { Argon2idPasswordHasher } from '../../../adapters/crypto/argon2id-password-hasher'
import { InMemoryPasswordHistoryRepository } from '../../../adapters/persistence/memory/password-history-repo'
import { InMemoryUserRepository } from '../../../adapters/persistence/memory/user-repo'
import { type DomainEvent, UserSchema } from '../../spec-bindings/schemas'

import {
  type AdminUserDeps,
  UsernameConflictError,
  UserNotFoundError,
  createAdminUser,
  setAdminUserDisabled,
  updateAdminUser,
} from './users'

const HASHER = new Argon2idPasswordHasher(8, 1)

async function setup(): Promise<{
  deps: AdminUserDeps
  userRepo: InMemoryUserRepository
  events: DomainEvent[]
}> {
  const userRepo = new InMemoryUserRepository()
  const passwordHistoryRepo = new InMemoryPasswordHistoryRepository()
  const events: DomainEvent[] = []
  await userRepo.save(
    UserSchema.parse({
      sub: 'user-bob',
      tenant_id: 'default',
      preferred_username: 'bob',
      password_hash: await HASHER.hash('bob-password-12345'),
      email: 'bob@example.com',
      email_verified: false,
      mfa_enrolled: false,
      roles: [],
      created_at: '2024-01-01T00:00:00.000Z',
      updated_at: '2024-01-01T00:00:00.000Z',
    }),
  )
  const deps: AdminUserDeps = {
    userRepo,
    passwordHasher: HASHER,
    passwordHistoryRepo,
    emit: (e) => events.push(e),
  }
  return { deps, userRepo, events }
}

describe('createAdminUser', () => {
  it('roles を normalize して保存し UserCreated を emit', async () => {
    const { deps, userRepo, events } = await setup()
    const user = await createAdminUser(deps, {
      actorSub: 'user-admin',
      tenant_id: 'default',
      preferred_username: 'carol',
      password: 'fresh-password-12345',
      email: 'carol@example.com',
      roles: ['support', 'admin', 'support'],
    })
    expect(user.roles).toEqual(['admin', 'support'])
    expect((await userRepo.findByUsername('default', 'carol'))?.sub).toBe(user.sub)
    expect(events.some((e) => e.type === 'UserCreated')).toBe(true)
  })

  it('username 衝突は UsernameConflictError', async () => {
    const { deps } = await setup()
    await expect(
      createAdminUser(deps, {
        actorSub: 'user-admin',
        tenant_id: 'default',
        preferred_username: 'bob',
        password: 'fresh-password-12345',
      }),
    ).rejects.toBeInstanceOf(UsernameConflictError)
  })
})

describe('updateAdminUser', () => {
  it('変更なしなら何も emit しない', async () => {
    const { deps, events } = await setup()
    await updateAdminUser(deps, { actorSub: 'user-admin', sub: 'user-bob' })
    expect(events.filter((e) => e.type === 'UserUpdated')).toHaveLength(0)
  })

  it('email_verified を true に変更すると changedFields に乗る', async () => {
    const { deps, events, userRepo } = await setup()
    await updateAdminUser(deps, {
      actorSub: 'user-admin',
      sub: 'user-bob',
      email_verified: true,
    })
    expect((await userRepo.findBySub('user-bob'))?.email_verified).toBe(true)
    const event = events.find((e) => e.type === 'UserUpdated') as
      | { changedFields: string[] }
      | undefined
    expect(event?.changedFields).toEqual(['email_verified'])
  })

  it('存在しない sub は UserNotFoundError', async () => {
    const { deps } = await setup()
    await expect(
      updateAdminUser(deps, { actorSub: 'user-admin', sub: 'nope', name: 'X' }),
    ).rejects.toBeInstanceOf(UserNotFoundError)
  })
})

describe('setAdminUserDisabled', () => {
  it('disable→enable で UserDisabled / UserEnabled を順に emit', async () => {
    const { deps, events, userRepo } = await setup()
    await setAdminUserDisabled(deps, { actorSub: 'user-admin', sub: 'user-bob', disabled: true })
    expect((await userRepo.findBySub('user-bob'))?.disabled_at).toBeDefined()
    await setAdminUserDisabled(deps, { actorSub: 'user-admin', sub: 'user-bob', disabled: false })
    expect((await userRepo.findBySub('user-bob'))?.disabled_at).toBeUndefined()
    expect(events.map((e) => e.type)).toContain('UserDisabled')
    expect(events.map((e) => e.type)).toContain('UserEnabled')
  })

  it('既に disabled の user を再び disable しても idempotent', async () => {
    const { deps, events, userRepo } = await setup()
    await setAdminUserDisabled(deps, { actorSub: 'user-admin', sub: 'user-bob', disabled: true })
    const before = (await userRepo.findBySub('user-bob'))?.disabled_at
    const eventsBefore = events.length
    await setAdminUserDisabled(deps, { actorSub: 'user-admin', sub: 'user-bob', disabled: true })
    const after = (await userRepo.findBySub('user-bob'))?.disabled_at
    expect(after).toBe(before!)
    expect(events.length).toBe(eventsBefore)
  })
})

File: ra-idp-go/internal/administration/usecases/users_test.go
package usecases_test

import (
	"context"
	"testing"
	"time"

	"ra-idp-go/internal/adapters/crypto"
	"ra-idp-go/internal/adapters/persistence/memory"
	adminusecases "ra-idp-go/internal/administration/usecases"
	"ra-idp-go/internal/spec"
)

func TestCreateUpdateAndDisableUser(t *testing.T) {
	ctx := context.Background()
	userRepo := memory.NewUserRepository()
	historyRepo := memory.NewPasswordHistoryRepository()
	hasher := crypto.NewArgon2idPasswordHasher()
	var events []spec.DomainEvent
	deps := adminusecases.Deps{
		UserRepo: userRepo, PasswordHasher: hasher, PasswordHistoryRepo: historyRepo,
		Emit: func(event spec.DomainEvent) { events = append(events, event) },
	}
	now := time.Date(2026, 6, 13, 12, 0, 0, 0, time.UTC)
	email := "bob@example.com"
	user, err := adminusecases.CreateUser(ctx, deps, adminusecases.CreateUserInput{
		ActorSub: "admin", PreferredUsername: "bob", Password: "initial-password-9182",
		Email: &email, Roles: []string{"support", "support"}, Now: now,
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(user.Roles) != 1 || user.Roles[0] != "support" {
		t.Fatalf("roles=%v", user.Roles)
	}
	if events[0].EventType() != "UserCreated" {
		t.Fatalf("event=%s", events[0].EventType())
	}
	updatedName := "Bob"
	roles := []string{"admin", "support"}
	user, err = adminusecases.UpdateUser(ctx, deps, adminusecases.UpdateUserInput{
		ActorSub: "admin", Sub: user.Sub, Name: &updatedName, Roles: &roles, Now: now.Add(time.Minute),
	})
	if err != nil {
		t.Fatal(err)
	}
	if user.Name == nil || *user.Name != "Bob" || len(user.Roles) != 2 {
		t.Fatalf("updated user=%+v", user)
	}
	user, err = adminusecases.SetUserDisabled(
		ctx, deps, "admin", user.Sub, true, now.Add(2*time.Minute),
	)
	if err != nil {
		t.Fatal(err)
	}
	if user.DisabledAt == nil {
		t.Fatal("disabled_at was not set")
	}
	if got := events[len(events)-1].EventType(); got != "UserDisabled" {
		t.Fatalf("last event=%s", got)
	}
	user, err = adminusecases.SetUserDisabled(
		ctx, deps, "admin", user.Sub, false, now.Add(3*time.Minute),
	)
	if err != nil {
		t.Fatal(err)
	}
	if user.DisabledAt != nil {
		t.Fatal("disabled_at was not cleared")
	}
	if got := events[len(events)-1].EventType(); got != "UserEnabled" {
		t.Fatalf("last event=%s", got)
	}
}

func TestCreateUserRejectsDuplicateUsername(t *testing.T) {
	repo := memory.NewUserRepository()
	now := time.Now().UTC()
	repo.Seed(&spec.User{
		Sub: "existing", PreferredUsername: "bob", PasswordHash: "hash",
		CreatedAt: now, UpdatedAt: now,
	})
	_, err := adminusecases.CreateUser(context.Background(), adminusecases.Deps{
		UserRepo: repo, PasswordHasher: crypto.NewArgon2idPasswordHasher(),
		PasswordHistoryRepo: memory.NewPasswordHistoryRepository(),
	}, adminusecases.CreateUserInput{
		PreferredUsername: "bob", Password: "initial-password-9182",
	})
	if err != adminusecases.ErrUsernameConflict {
		t.Fatalf("error=%v, want ErrUsernameConflict", err)
	}
}
